internal/datacoord/meta.go

```diff
@@ -1366,92 +1387,85 @@ func (m *meta) completeClusterCompactionMutation(t *datapb.CompactionTask, resul
 			NumOfRows:           seg.NumOfRows,
 			State:               commonpb.SegmentState_Flushed,
-			MaxRowNum:           latestCompactFromSegments[0].MaxRowNum,
+			MaxRowNum:           compactFromSegInfos[0].MaxRowNum,
 			Binlogs:             seg.GetInsertLogs(),
 			Statslogs:           seg.GetField2StatslogPaths(),
 			CreatedByCompaction: true,
 			CompactionFrom:      compactFromSegIDs,
 			LastExpireTime:      tsoutil.ComposeTSByTime(time.Unix(t.GetStartTime(), 0), 0),
 			Level:               datapb.SegmentLevel_L2,
-			StartPosition: getMinPosition(lo.Map(latestCompactFromSegments, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
+			StartPosition: getMinPosition(lo.Map(compactFromSegInfos, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
 				return info.GetStartPosition()
 			})),
-			DmlPosition: getMinPosition(lo.Map(latestCompactFromSegments, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
+			DmlPosition: getMinPosition(lo.Map(compactFromSegInfos, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
 				return info.GetDmlPosition()
 			})),
 		}
 		segment := NewSegmentInfo(segmentInfo)
-		newSegments = append(newSegments, segment)
+		compactToSegInfos = append(compactToSegInfos, segment)
+		compactToSegIDs = append(compactToSegIDs, segment.GetID())
 		metricMutation.addNewSeg(segment.GetState(), segment.GetLevel(), segment.GetNumOfRows())
 	}
-	compactionTo := make([]UniqueID, 0, len(newSegments))
-	for _, s := range newSegments {
-		compactionTo = append(compactionTo, s.GetID())
-	}
 
-	log.Info("meta update: prepare for complete compaction mutation - complete",
-		zap.Int64("collectionID", latestCompactFromSegments[0].CollectionID),
-		zap.Int64("partitionID", latestCompactFromSegments[0].PartitionID),
-		zap.Any("compacted from", compactFromSegIDs),
-		zap.Any("compacted to", compactionTo))
+	log = log.With(zap.Int64s("compact from", compactFromSegIDs), zap.Int64s("compact to", compactToSegIDs))
+	log.Debug("meta update: prepare for meta mutation - complete")
 
-	compactFromInfos := lo.Map(latestCompactFromSegments, func(info *SegmentInfo, _ int) *datapb.SegmentInfo {
+	compactFromInfos := lo.Map(compactFromSegInfos, func(info *SegmentInfo, _ int) *datapb.SegmentInfo {
 		return info.SegmentInfo
 	})
 
-	newSegmentInfos := lo.Map(newSegments, func(info *SegmentInfo, _ int) *datapb.SegmentInfo {
+	compactToInfos := lo.Map(compactToSegInfos, func(info *SegmentInfo, _ int) *datapb.SegmentInfo {
 		return info.SegmentInfo
 	})
 
 	binlogs := make([]metastore.BinlogsIncrement, 0)
-	for _, seg := range newSegmentInfos {
+	for _, seg := range compactToInfos {
 		binlogs = append(binlogs, metastore.BinlogsIncrement{Segment: seg})
 	}
-	if err := m.catalog.AlterSegments(m.ctx, append(compactFromInfos, newSegmentInfos...), binlogs...); err != nil {
-		log.Warn("fail to alter segments and new segment", zap.Error(err))
+	// alter compactTo before compactFrom segments to avoid data lost if service crash during AlterSegments
```

Comment:
should we do 	m.segments.SetSegment(compactToSegmentInfo.GetID(), compactToSegmentInfo)
after we store compactToInfos?

Reply:
can not finish this call in one request seems to be dangerous.
can we only handle 100 segments max at a time?